### PR TITLE
RUN-2221: Update webpack 4 to 5

### PR DESCRIPTION
**Pull request by smartinellibenedetti**
**Is this a bugfix, or an enhancement? Please describe.**
-

List of enhancement/maintanance:

- Update webpack 4 to 5;
- Update packages related to vue-cli and jest;
- Remove/fix/update imports that had issues - some were wrong, some aren't needed anymore;
- Remove commands that were necessary only for webpack 4;
- Small fixes (especially syntax) that would prevent builds or tests from running;

**Describe the solution you've implemented**
-

Followed the guide https://webpack.js.org/migrate/5/  and the following order for updating:

updated webpack > other npm packages > errors that would prevent the project from running/building locally > tests /imports/syntax errors 

Checks performed:
- ran final image with docker;
- ran locally all commands used by the project;

How to test:
- 

- run this branch locally, firstly npm install then run project's `core` settings (can also do `enterprise` as well, but then please remember to run installUiTrellis first);
- check for any missing files or errors in the UI;

Or test with a docker image;

### **Describe alternatives you've considered**

In the past there have been efforts to move the FE to vite, but as it requires a bigger effort to move out from vue-cli+webpack to vite, updating webpack to the newest version was a compromise for improving build speed/security and also ensuring compatibility with new npm packages that require webpack 5.

**Additional context**

Added comments explaining what was done in this PR. 

Also, it might be worth to consider to use other modes of devtools as it greatly improves build time (One run, which had other modes enabled had a performance gain of 40% https://app.circleci.com/pipelines/github/rundeck/rundeck/14735)

**_Final performance improvement on circle ci: 5%_**

### CODE REVIEW

File: rundeckapp/grails-spa/packages/ui-trellis/src/library/utilities/uivi18n.ts [hunk 1]
```diff
@@ -1,60 +1,21 @@
...
-  ru_RU,
-  ru: ru_RU,
-  sv_SE,
-  sv: sv_SE,
-  tr_TR,
-  tr: tr_TR,
-  zh_CN,
-  zh: zh_CN,
-};
+export default {};
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/src/library/utilities/uivi18n.ts:21]: uiv doesn't export those anymore - will possibly just delete the file once the rest is working

File: rundeckapp/grails-app/views/framework/_createProjectForm.gsp [hunk 2]
```diff
@@ -82,7 +82,6 @@
 
     <asset:javascript src="framework/editProject.js"/>
     <asset:javascript src="static/pages/project-config.js" defer="defer" />
-    <asset:stylesheet href="static/css/pages/project-config.css" />
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/framework/_createProjectForm.gsp:85]: This file isn't being created, as the respective components do not have any css to be imported.

File: rundeckapp/grails-app/views/framework/adhoc.gsp [hunk 3]
```diff
@@ -34,7 +34,7 @@
           kind: AuthConstants.TYPE_EVENT
   )}"/>
 
-    <asset:javascript src="static/css/pages/command.css"/>
+    <asset:stylesheet src="static/css/pages/command.css"/>
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/framework/adhoc.gsp:37]: This was causing a 404 as the generated import had the wrong path and extension

File: rundeckapp/grails-app/views/framework/editProject.gsp [hunk 4]
```diff
@@ -90,7 +90,6 @@
     <asset:javascript src="leavePageConfirm.js"/>
     <asset:javascript src="framework/editProject.js"/>
     <asset:javascript src="static/pages/project-config.js" defer="defer" />
-    <asset:stylesheet href="static/css/pages/project-config.css" />
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/framework/editProject.gsp:93]: This file isn't being created, as the respective components do not have any css to be imported.

File: rundeckapp/grails-app/views/menu/home.gsp [hunk 5]
```diff
@@ -62,9 +62,6 @@
 
 	<asset:javascript src="menu/home.js"/>
 
-	<!-- VUE CSS MODULES -->
-	<asset:stylesheet href="static/css/components/version-notification.css"/>
-	<!-- /VUE CSS MODULES -->
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/menu/home.gsp:67]: This file isn't being created. In fact the components in the folder version-notification do not seem to be used anywhere.

File: rundeckapp/grails-app/views/menu/home.gsp [hunk 6]
```diff
@@ -462,9 +459,7 @@
 	</div>
 	<!-- VUE JS MODULES -->
 	<asset:stylesheet href="static/css/pages/home.css"/>
-	<asset:stylesheet href="static/css/components/first-run.css"/>
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/menu/home.gsp:465]: This file isn't being created, but the import isn't needed as the CSS required by its components is included in the chunk-common.css

File: rundeckapp/grails-app/views/menu/home.next.gsp [hunk 7]
```diff
@@ -63,7 +63,6 @@
   <asset:javascript src="menu/home.js"/>
 
   <!-- VUE CSS MODULES -->
-  <asset:stylesheet href="static/css/components/version-notification.css"/>
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/menu/home.next.gsp:66]: This file isn't being created. In fact the components in the folder version-notification do not seem to be used anywhere.

File: rundeckapp/grails-app/views/menu/home.next.gsp [hunk 8]
```diff
@@ -80,7 +79,7 @@
   <div id="layoutBody">
     <div class="vue-ui-socket">
       <g:set var="createProjectAllowed"
-             value="${auth.resourceAllowedTest(action: AuthConstants.ACTION_CREATE, type: AuthConstants.TYPE_PROJECT, context: AuthConstants.CTX_APPLICATION)}"/>
+             value="${auth.resourceAllowedTest(action: [AuthConstants.ACTION_CREATE], type: AuthConstants.TYPE_PROJECT, context: AuthConstants.CTX_APPLICATION, has:true)}"/>
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/menu/home.next.gsp:82]: home.next is behind a feature flag but noticed that this line wasn't 100% its equivalent in home.gsp

File: rundeckapp/grails-app/views/menu/home.next.gsp [hunk 9]
```diff
@@ -104,7 +103,6 @@
   </div>
   <!-- VUE JS MODULES -->
   <asset:stylesheet href="static/css/pages/home.css"/>
-  <asset:javascript src="static/components/version-notification.js"/>
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-app/views/menu/home.next.gsp:107]: This file isn't being created. In fact the components in the folder version-notification do not seem to be used anywhere.

File: rundeckapp/grails-spa/packages/ui-trellis/.env.development [hunk 10]
```diff
@@ -1,3 +1,3 @@
 VUE_APP_OUTPUT_DIR=../../../grails-app/assets/provided/static
-VUE_APP_DEVTOOL=eval-cheap-module-source-map
+VUE_APP_DEVTOOL=eval-cheap-source-map
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/.env.development:2]: eval-cheap-source-map has better performance than eval-cheap-module-source-map. Open to discussion if team would like to try faster modes:

https://webpack.js.org/configuration/devtool/#devtool

File: rundeckapp/grails-spa/packages/ui-trellis/jest.config.js [hunk 11]
```diff
@@ -1,13 +1,8 @@
 const esModules = ["vue-virtual-scroller", "uuid"].join("|");
 module.exports = {
-  globals: {
-    "ts-jest": {
-      tsconfig: "tsconfig.app.json",
-    },
-  },
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/jest.config.js:7]: outdated format

File: rundeckapp/grails-spa/packages/ui-trellis/package.json [hunk 12]
```diff
@@ -13,12 +13,12 @@
     "ci:app:build": "npm ci && npm run build:app",
     "dev:library:build": "npm install && npm run build:library --prefer-offline --no-audit",
     "ci:library:build": "npm ci && npm run build:library --prefer-offline --no-audit",
-    "dev": "export NODE_OPTIONS=--openssl-legacy-provider && VUE_CLI_SERVICE_CONFIG_PATH=$PWD/vue.config.app.js ./node_modules/.bin/vue-cli-service build --mode development --watch && tsc -p ./tsconfig.app.json && VUE_CLI_SERVICE_CONFIG_PATH=$PWD/vue.config.library.js ./node_modules/.bin/vue-cli-service build --mode development --watch",
+    "dev": "VUE_CLI_SERVICE_CONFIG_PATH=$PWD/vue.config.app.js ./node_modules/.bin/vue-cli-service build --mode development --watch && tsc -p ./tsconfig.app.json && VUE_CLI_SERVICE_CONFIG_PATH=$PWD/vue.config.library.js ./node_modules/.bin/vue-cli-service build --mode development --watch",
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/package.json:16]: `export NODE_OPTIONS=--openssl-legacy-provider` with the update this flag isn't necessary anymore

File: rundeckapp/grails-spa/packages/ui-trellis/src/app/components/home/tests/HomeCardList.spec.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import { shallowMount } from "@vue/test-utils";
+import { shallowMount, VueWrapper } from "@vue/test-utils";
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/src/app/components/home/tests/HomeCardList.spec.ts:1]: IIRC correctly, these changes (both in HomeCardList and HomeView) were merged to main with another PR, but somehow when I applied the rebase they ended up becoming a separate commit.

File: rundeckapp/grails-spa/packages/ui-trellis/src/library/components/plugins/pluginPropView.vue [hunk 14]
```diff
@@ -3,51 +3,49 @@
     <span v-if="prop.type === 'Boolean'" class="configpair">
       <template
         v-if="
-          value === 'true' ||
-          value === true ||
-          (prop.defaultValue === 'true' &&
-            (value === 'false' || value === false))
+          innerValue === 'true' ||
+          (prop.defaultValue === 'true' && innerValue === 'false')
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/src/library/components/plugins/pluginPropView.vue:7]: Fixed vue issues. I didn't add tests to limit the scope of the PR.

File: rundeckapp/grails-spa/packages/ui-trellis/vue.config.library.js [hunk 15]
```diff
@@ -62,21 +61,27 @@ module.exports = {
...
-        return "[name].vue.js";
-      else return "[name].js";
+      // assets at this point are either test files or vue components
+      // previously webpack was including the extension in the name
+      // now it doesn't do so for the vue files, so asset.chunk.name will
+      // either be the component name, or component name.test
+      if (asset.chunk.name.endsWith("test")) return "[name].js";
+      else return "[name].vue.js";
     };
-    config.output.library = "rundeckUiTrellis";
```
Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/vue.config.library.js:72]: Removed, in accordance to: https://github.com/webpack/webpack/issues/11800

Comment (smartinellibenedetti) [inline rundeckapp/grails-spa/packages/ui-trellis/vue.config.library.js:72]: Removed as libraryTarget with value of commonjs2 doesn't work with `library` property

File: rundeckapp/grails-app/views/menu/home.gsp [hunk 16]
```diff
@@ -462,9 +459,7 @@
...
 	<asset:javascript src="static/components/first-run.js"/>
-	<asset:javascript src="static/components/version-notification.js"/>
```
[the remark this answers is not in the document]
Comment (gschueler) [reply rundeckapp/grails-app/views/menu/home.gsp:467]: seems so, let's do a separate PR to remove that code since it isn't used

**Review by gschueler — APPROVED** [review]
LGTM, did a spot check on a variety of the pages that might have been affected, but they all look good

File: rundeckapp/grails-app/views/menu/home.gsp (hunk 16 above)
[the remark this answers is not in the document]
Comment (smartinellibenedetti) [reply rundeckapp/grails-app/views/menu/home.gsp:467]: Okay, PR for delete: https://github.com/rundeck/rundeck/pull/9135

**Pull request merged**